### ISSUE TITLE
Options for portability of PERCEIVE Hadoop Ecosystem

### ISSUE THREAD

**Issue by yueliu2368** [sailuh/perceive#53]
I read the book about using Vagrant, Vagrant: up and Running.  

I also investigated several ways to build a virtual machine with Vagrant and auto-provisioning using shell script and puppet. 

Out of the available options, It is more convenient for me to use vagrant to build a barebone Ubuntu server VM, connect the VM and use linux code to install NIFI, Hadoop, etc , and pack the existing environment as box, instead of using configuration tools like puppet and shell scripts to realize auto-provisioning. 

However, the box at the minimum is a couple hundred megabytes or a few gigabyes. Therefore, I want to confirm that whether this kind of box is acceptable for our project and meet the relevant requirements. 

If it is acceptable, I will finish the installation of NIFI using vagrant before the Sunday and investigate how to install Hadoop for next week.

**Comment by carlosparadis** [comment]
Why would the box be already a few gigabytes? If I recall correctly, vagrant allows you to configure it so it downloads when you deploy it. Maybe we could arrange in this way for the installation at least for the OS? On the other hand side, we may have problems with ensuring it is portable enough. Thoughts?

**Comment by yueliu2368** [comment]
Thanks for your reply. If I understand correctly, do you mean I should adopt the same way as the following link? https://github.com/DandyDev/virtual-hadoop-cluster

**Comment by carlosparadis** [comment]
That is a really good finding, and yes, it is what I meant. A minimal file with the necessary setup to deploy the ecosystem which is easy to share and version. 

I don't object to using CDH5 to then sub-select what packages we want but take note we may not be able to obtain all projects we would like to have in one of these Hadoop ecosystems managers. 

Why don't you take a look and see if it offers what we considered starting with? Also, may want to check https://hortonworks.com/ to compare against Cloudera. Keep a close eye to Apache Kite or Apache NiFi, and of course, anything that can substitute them.

**Comment by yueliu2368** [comment]
Great! I will finish the setup file for NIFI before this Sunday.

**Comment by carlosparadis** [comment]
Hi, Yue, how is it coming along?

**Comment by yueliu2368** [comment]
Sorry for the late response. I have finished the configuration shell scripts for NIFI service on the last Sunday. Shall I post the code on GitHub or wait until I finish all configuration scripts for NIFI, Hadoop and Kite? For this week task, I will review the content you presented last Saturday, check into the MEGA account you posted, and investigate some github repositories that may inspire me (https://github.com/minyk/nifi-sandbox). Since there are a lot of SH files, I probably will enroll some revelent course about shell scrips and edit the code based on that repository and our project.

**Comment by carlosparadis** [comment]
Before you go ahead and submit anything, why don't you try to give me a big picture of what you did, and then we decide what is necessary or not? Have you taken the Cloudera CDH5 route in the end? Where the configuration scripts fit in? Feel free to add screenshots here (just don't send them on Pull Requests unless previously agreed it is needed :-) ).

P.s.: As you go, keep track of your links here, it will be useful for the group!

**Comment by yueliu2368** [comment]
After making some research and comparison between Cloudera CDH, Hortonworks product and Apache Hadoop, my conclusion is that Apache Hadoop is more suitable for overall project. Here is what I am thinking: based on your clarification under Tu’s issue, data storage under big data platform and data flow management are more important based on our goal. Although Hortonworks products, HDP and HDF, already include NIFI and apache Hadoop, a lot of components provided by Hortonworks may not be used for the overall projects and thus waste the server resources, like Storm, Kafka. Therefore, I think the combination of Apache Hadoop 2.8.3, Apache NIFI and Apache Kite should be enough for our mission.  As a result, I plan to use Vagrant to build configuration scripts running on Centos 7 OS pure box that automatically install the above three tools .

**Comment by carlosparadis** [comment]
@yueliu2368 I think they allow you to install only what is necessary rather than the entire thing. However, as you saw, they do have divergence on products. For example, I think Apache Storm comes in `Northonworks`, but it is not part of `Cloudera` (correct if I am wrong). 

Could you double check this? My main concern comes when we see a need of a new project that is not part of their suite: Now your overhead is not just to understand integration between them, but also between them and Cloudera, which may be too hard to deal with (or easier, I don't know).

**Comment by carlosparadis** [comment]
@yueliu2368 So we have a starting point for a deliverable, could you make a pull request with a minimal vagrant virtual box configured to install a Ubuntu OS? You can put it on `perceive/Database/Hadoop_Ecosystem`. 

The database folder is a bit messy but I will clean it up soon (and in the process, I may create the folder too, but if you don't find it feel free to create it).

**Comment by yueliu2368** [comment]
Sure, I will make a pull requrest with a vagrant virtual box configuration scripts to install a Ubuntu OS and NIFI.

**Comment by carlosparadis** [comment]
@yueliu2368 whatever you do, just don't upload a 4GB virtual box machine as a pull request! Haha :-) Just configs please! I would imagine this would be a few KB but nothing else. Github will come after us if you even dream of uploading a large blob like that.

**Comment by yueliu2368** [comment]
Definitelly, haha. just the vagrantFile configuration scripts.

**Comment by carlosparadis** [comment]
As an update from our discussion:

The idea will be to:

 * Create an apache nifi + kyte + apache spark pipeline (see [this link](https://blogs.apache.org/nifi/entry/stream_processing_nifi_and_spark) for ideas on nifi + spark). 
* Get comfortable with adding connectors to ingest data from disk (your full disclosure files I sent you the link on mega), and the other from [capec (see the bottom of the page for the XML links)](capec.mitre.org/data). For the CAPEC, use a connector that downloads directly from the URL just for practice. 

Afterwards, try to connect to HDFS via terminal to see the files are indeed there. Try to also get familiar with how Nifi keeps track of what has been executed and when, as this is why we are trying to use NiFi in the first place: Data Provenance. 

Vagrant is a plus, but the data pipeline is the priority. 

As a next step, we can get you started on more on the analysis side of Spark by rewriting [this portion](https://github.com/sailuh/perceive/tree/master/Data_Collection/Full_Disclosure/Pre_Processing/Reply_To_Threads) in Apache Spark to group files into threads. We can start by applying to each file a bag of words transformation, and then reduce them by key (the e-mail title) by summing the words, as the bag of words representation is additive. More on this in a separate issue when we get there.

p.s.: Don't forget to fix that PR with the commit label! :)

**Comment by yueliu2368** [comment]
Here is the summary and update from our discussion: 
   
 The tutorial of this [link](https://blogs.apache.org/nifi/entry/stream_processing_nifi_and_spark) requires Java and maven to install dependency between Spark and nifi and also lack of the explanation about where to edit config and how. Without appropriate dependency between Spark and Nifi, we may need to re-evaluate how we integrate Apache Nifi and Apache Spark.

Since our data is not streaming in the true sense, it is OK to stay with batch rather than worry about 'streaming data'. As such, one viable option for us is to ignore Apache Spark talking to Apache nifi. Instead, we can have nifi keep an eye on the data all the way down to sitting in HDFS, and having the reshape pipeline also done there. However, if we do so, we may lose the big picture from NIFI. but better than nothing. 

The idea about my next step:
1. Build a Python notebook with the following questions:
- Is the raw data still there?
- Is the processed data now sitting in HDFS processed?
- How to add multiple files?
- If we do have multiple files, are we running in parallel?
- What connectors are we using? Why? How to configure them?
- What do I need to do to run?
- What configurations are needed in the connectors?
- What is going on after I click to execute the NiFi Pipeline?
- NiFi best selling point is data provenance. What can we gain by using that versus just doing everything with NiFi? What are the pros? What are the cons? Is it worthwhile the overhead of a new project?
2. In the above notebook, also include the tutorial and screenshots about setting up and basic usage of nifi
3. Adding Kite sdk connector in nifi and play with the conenctors that filter and split data
4. Try to build a dependency between Spark and nifi using Python
